settings: Auto-update JSON schemas for settings when extensions are un/installed

probably-neb opened the pull request:
Because of #26562, it is now possible to subscribe to extension update events within the LSP store, where we can then update the Schemas sent to the JSON LSP resulting in dynamic updates to the auto-complete suggestions and diagnostics in settings. Notably, this means newly installed languages and (icon) themes will auto-complete correctly as soon as the extension is installed.

Closes #15436

Release Notes:

- Fixed an issue where autocomplete suggestions and diagnostics for languages and (icon) themes in settings would not update when the extension with which they were added was installed or uninstalled

probably-neb committed:
impl

probably-neb committed:
fearless concurrency!

probably-neb committed:
clean + document

probably-neb committed:
remove extraneous Arc param

probably-neb committed:
forgot about headless project

probably-neb committed:
clippppppy

probably-neb committed:
clippppppy

probably-neb committed:
oops

maxdeviant commented on crates/extension_host/src/extension_host.rs:
```diff
@@ -1213,9 +1213,6 @@ impl ExtensionStore {
 
         self.extension_index = new_index;
         cx.notify();
-        ExtensionEvents::global(cx).update(cx, |this, cx| {
```
We use this event elsewhere to know when the set of extensions changed (but haven't necessarily fully reloaded), so I don't think we should change the location of where we emit it.

I think it would make sense to add a separate `ExtensionsReloaded` event if we care about knowing that they have been fully reloaded.

probably-neb replied:
Makes sense will do

maxdeviant commented on crates/remote_server/src/headless_project.rs:
```diff
@@ -139,13 +139,16 @@ impl HeadlessProject {
             observer
         });
 
+        let extension_events = extension::ExtensionEvents::global(cx);
```
Do we need to pass this in as a parameter, or could we retrieve it inside of the constructor since we have a `cx`?

probably-neb replied:
My assumption was that it wouldn't be needed in the headless host, so the thing I meant to do and then forgot about until the CI failed is pass in an option so that `new_local` knows whether to subscribe or not

probably-neb committed:
pass optional extension_events to `LspStore::new_local`

probably-neb committed:
Create new event rather than moving emit location of ExtensionsUpdated

Also: revert changes that were made to extract ExtensionsUpdated event, as they are no longer necessary

probably-neb committed:
use try_global rather than panicing when ExtensionEvents global is not initialized

probably-neb committed:
missed one
